components/payments/internal/storage/bank_accounts.go

```diff
@@ -38,30 +38,48 @@ type bankAccount struct {
 	RelatedAccounts []*bankAccountRelatedAccount `bun:"rel:has-many,join:id=bank_account_id"`
 }
 
-func (s *store) BankAccountsUpsert(ctx context.Context, bankAccount models.BankAccount) error {
+func (s *store) BankAccountsUpsert(ctx context.Context, ba models.BankAccount) error {
 	tx, err := s.db.BeginTx(ctx, nil)
 	if err != nil {
 		return e("begin transaction", err)
 	}
 	defer tx.Rollback()
 
-	toInsert := fromBankAccountModels(bankAccount)
+	toInsert := fromBankAccountModels(ba)
 	// Insert or update the bank account
-	_, err = tx.NewInsert().
+	var idsUpdated []uuid.UUID
+	err = tx.NewInsert().
 		Model(&toInsert).
+		Column("id", "created_at", "name", "country", "metadata").
 		On("CONFLICT (id) DO NOTHING").
-		Exec(ctx)
+		Returning("id").
+		Scan(ctx, &idsUpdated)
 	if err != nil {
 		return e("insert bank account", err)
 	}
 
-	// Insert or update the related accounts
-	_, err = tx.NewInsert().
-		Model(&toInsert.RelatedAccounts).
-		On("CONFLICT (bank_account_id, account_id) DO NOTHING").
-		Exec(ctx)
-	if err != nil {
-		return e("insert related accounts", err)
+	if len(idsUpdated) > 0 {
+		_, err = tx.NewUpdate().
+			Model((*bankAccount)(nil)).
+			Set("account_number = pgp_sym_encrypt(?::TEXT, ?, ?)", toInsert.AccountNumber, s.configEncryptionKey, encryptionOptions).
+			Set("iban = pgp_sym_encrypt(?::TEXT, ?, ?)", toInsert.IBAN, s.configEncryptionKey, encryptionOptions).
+			Set("swift_bic_code = pgp_sym_encrypt(?::TEXT, ?, ?)", toInsert.SwiftBicCode, s.configEncryptionKey, encryptionOptions).
+			Where("id = ?", toInsert.ID).
```

Comment:
On update, you should always update fields ONLY if it changed.
Since the function is an upsert, I guess account_number, iban, and swift_bic_code can already have the correct value.
So you should check for that in the where clause.
If you do not, the values will be rewritten even if it does not change. And it will involve all the write process of postgres, including reindexing.

Reply:
Discussed IRL, we update only when the bank account is created ! We have some tests on that, so everything is good !